pyquil/paulis.py

```diff
@@ -677,6 +692,83 @@ def get_programs(self):
         coefficients = np.array([term.coefficient for term in self.terms])
         return programs, coefficients
 
+    def matrix(self, qubit_mapping={}, n_qubits=None):
+        """Create the matrix representation of self.
+
+        :param n_qubits:  Number of qubits to represent self on. If None, the maximum qubit from
+                         the PauliSum or qubit_mapping is used
+        :param qubit_mapping: A dictionary-like object that maps from :py:class`QubitPlaceholder`
+                              to :py:class:`int`
+        :return: np.matrix representing the PauliTerm
+        """
+
+        # get a list of the qubits in self
+        self_qubits = self.get_qubits()
+        # find maximum real qubit in self for later use
+        max_qubit = 0
+        for qubit in self_qubits:
+            if isinstance(qubit, int):
+                if qubit > max_qubit:
+                    max_qubit = qubit
+
+        # n_qubits defaults to the maximum of the mapped or used qubits
+        # (depending on which is larger)
+        if n_qubits is None:
+            n_qubits = max({max_qubit} | {*qubit_mapping.values()}) + 1
+        elif n_qubits < max({max_qubit} | {*qubit_mapping.values()}) + 1:
+            raise ValueError("n_qubits is smaller than the highest qubit in"
+                             " the hamiltonian")
+
+        # create a list of qubits ordered by how they should appear in the
+        # tensor product.
+        # We don't automatically skip in all real qubits, because  we want to
+        # be able to also map real qubits to other indices in the tensor
+        # product
+        qubit_list = [*range(n_qubits)]
```

Comment:
The logic above + this line relates to the problem I had brought up before. Let's say we have `pt = sZ(0) * sX(2)` and `qubit_mapping = {q0: 0, q1: 2}`, then this logic seems to identify that there are 3 qubits, and would produce a 8x8 matrix representation (even though there are only 2 qubits, and it should produce a 4x4 matrix representation).

Reply:
You are right! It took me way too long, to realize that X0*X2 should (by default) be represented on 2 qubits and not 3. The Wavefunction Simulator also returns a wavefunction with only 4 entries instead of 8, if I have e.g. p = Program(X(0), H(2)). It is fixed now.